bot/cogs/eval.py

```diff
@@ -12,66 +19,153 @@
-        """
-        Bot owner only: Evaluate Python code
+        res = ""
 
-        Your code may be surrounded in a code fence, but it's not required.
-        Scope will be preserved - variables set will be present later on.
-        """
+        # Erase temp input we made
+        if inp.startswith("_ = "):
+            inp = inp[4:]
 
-        code = string.strip()
+        lines = [l for l in inp.split("\n") if l.strip()]
+        if len(lines) != 1:
+            lines += [""]
+
+        # Create the inpit dialog
+        for i, line in enumerate(lines):
+            if i == 0:
+                s = f"In [{self.ln}]: "
 
-        if code.startswith("```") and code.endswith("```"):
-            if code.startswith("```python"):
-                code = code[9:-3]
-            elif code.startswith("```py"):
-                code = code[5:-3]
             else:
-                code = code[3:-3]
-        elif code.startswith("`") and code.endswith("`"):
-            code = code[1:-1]
+                # Indent the 3 dots correctly
+                s = (f"{{:<{len(str(self.ln))+2}}}...: ").format("")
 
-        code = code.strip().strip("\n")
-        io = StringIO()
+            if i == len(lines) - 2:
+                if line.startswith("return"):
+                    line = line[6:].strip()
+
+            res += s + line + "\n"
+
+        self.stdout.seek(0)
+        text = self.stdout.read()
+        self.stdout.close()
+        self.stdout = io.StringIO()
+
+        if text:
+            res += text + "\n"
+
+        if out is None:
+            # No output, return the input statement
+            return (res, None)
+
+        res += f"Out[{self.ln}]: "
+
+        if isinstance(out, discord.Embed):
+            # We made an embed? Send that as embed
+            res += "<Embed>"
+            res = (res, out)
 
-        try:
-            rvalue = await self.interpreter.run(code, ctx, io)
-        except Exception as e:
-            await ctx.send(
-                f"{ctx.author.mention} **Code**\n"
-                f"```py\n{code}```\n\n"
-                f"**Error**\n```{e}```"
-            )
         else:
-            out_message = (
-                f"{ctx.author.mention} **Code**\n"
-                f"```py\n{code}\n```"
-            )
-
-            output = io.getvalue()
-
-            if output:
-                out_message = (
-                    f"{out_message}\n\n"
-                    f"**Output**\n```{output}```"
-                )
-
-            if rvalue is not None:
-                out_message = (
-                    f"{out_message}\n\n"
-                    f"**Returned**\n```py\n{repr(rvalue)}\n```"
-                )
-
-            await ctx.send(out_message)
+            if (isinstance(out, str) and
+                    out.startswith("Traceback (most recent call last):\n")):
+                # Leave out the traceback message
+                out = "\n" + "\n".join(out.split("\n")[1:])
```

Comment:
it's got string concatenation, a .split nested inside a .join, and a container slice, all in one line. it's _too dense_. it'd be better to split this into more lines if it can't be simplified.